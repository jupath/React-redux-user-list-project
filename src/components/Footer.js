import React from 'react';

const Footer = () => (
  <footer className="footer pb-4 text-center">
    <h5>View <a href="https://github.com/jupath/React-redux-user-list-project/" target="_blank" rel="noopener noreferrer" className="font-weight-bold">Source on GitHub</a></h5>
  </footer>
);

export default Footer;
