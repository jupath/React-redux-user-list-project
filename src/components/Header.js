import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button, Input } from 'reactstrap';
import { setOrderBy, setText } from '../actions/filters';

export class Header extends Component {
  handleOrderByClick = (orderBy) => {
    this.props.setOrderBy(orderBy);
  }

  handleSetText = (event) => {
    const text = event.target.value;
    this.props.setText(text);
  }

  render() {
    return (
      <header className="header">
        <div className="container text-center py-4">
          <h1>User list built with React and Redux</h1>
          <p>This app uses async-await to fetch users from <a href="https://randomuser.me/" target="_blank">randomuser.me</a> API</p>
          <div className="px-lg-7 pb-4">
            <Input type="text" placeholder="Search by name..." className="header__searchform" onChange={this.handleSetText} />
          </div>
          <div>
            <span>Order list by name: </span>
            <Button id="ascButton" color="success" className="mr-1" disabled={this.props.orderBy === 'asc'} onClick={() => this.handleOrderByClick('asc')}>Asc</Button>
            <Button id="descButton" color="success" className="mr-1" disabled={this.props.orderBy === 'desc'} onClick={() => this.handleOrderByClick('desc')}>Desc</Button>
            <Button id="defaultButton" color="success" disabled={this.props.orderBy === undefined} onClick={() => this.handleOrderByClick()}>Default</Button>
          </div>
        </div>
      </header>
    );
  }
}

const mapStateToProps = state => ({
  orderBy: state.filters.orderBy,
});

const mapDispatchToProps = dispatch => ({
  setOrderBy: orderBy => dispatch(setOrderBy(orderBy)),
  setText: text => dispatch(setText(text)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
