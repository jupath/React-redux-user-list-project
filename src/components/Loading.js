import React from 'react';
import loader from '../images/spin.svg';

const Loading = () => (
  <div className="loading">
    <img src={loader} />
  </div>
);

export default Loading;