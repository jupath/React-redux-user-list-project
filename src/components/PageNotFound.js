import React from 'react';

const PageNotFound = () => (
  <div className="text-center pt-5">
    <h2>404 error!</h2>
    <p>Page not found!</p>
  </div>
);

export default PageNotFound;
