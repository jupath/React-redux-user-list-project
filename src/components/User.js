import React from 'react';
import Parser from 'html-react-parser';
import { Link } from 'react-router-dom';

const User = (props) => {
  const {
    lastName, firstName, image, username, cell,
  } = props.userData;

  return (
    <div className="col-sm-4 mb-4">
      <div className="user p-3">
        <div>
          <img src={image} className="mr-2" />
        </div>
        <div className="user__text">
          <h5 className="text-capitalize font-weight-bold">{Parser(lastName)}, {Parser(firstName)}</h5>
          <p>Cell: {cell}</p>
          <div className="user__text__btnwrapper">
            <Link className="btn btn-success" to={`/user/${username}`}>Details</Link>
          </div>
        </div>
      </div>
    </div>
  );
};

export default User;
