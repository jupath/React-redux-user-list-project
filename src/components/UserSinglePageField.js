import React, { Component } from 'react';
import { connect } from 'react-redux';
import FaPencil from 'react-icons/lib/fa/pencil';
import PropTypes from 'prop-types';
import UserSinglePageForm from './UserSinglePageForm';
import { updateUser } from '../actions/users';

export class UserSinglePageField extends Component {
  state = {
    isUpdate: false,
  }

  onUpdateButton = () => {
    this.setState({
      isUpdate: true,
    });
  }

  handleUpdateUser = (updatedData) => {
    this.props.updateUser(this.props.username, { [this.props.field]: updatedData });
    this.setState({
      isUpdate: false,
    });
  }

  render() {
    let renderField;
    if (this.state.isUpdate) {
      renderField = (<UserSinglePageForm
        data={this.props.data}
        handleUpdateUser={this.handleUpdateUser}
      />);
    } else {
      renderField = (<span className="user-single__text">
        {this.props.data}
        {this.props.isEditable &&
        <a
          role="button"
          tabIndex={0}
          className="ml-2"
          onClick={this.onUpdateButton}
          onKeyDown={this.onUpdateButton}
        >
          <FaPencil />
        </a>}
      </span>);
    }

    return (
      <div>
        { renderField }
      </div>
    );
  }
}

UserSinglePageField.propTypes = {
  updateUser: PropTypes.func.isRequired,
  username: PropTypes.string.isRequired,
  field: PropTypes.string.isRequired,
  data: PropTypes.string.isRequired,
  isEditable: PropTypes.bool.isRequired,
};

const mapDispatchToProps = dispatch => ({
  updateUser: (username, user) => dispatch(updateUser(username, user)),
});

export default connect(null, mapDispatchToProps)(UserSinglePageField);
