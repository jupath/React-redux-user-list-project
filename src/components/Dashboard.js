import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import filterUsers from '../selectors/users';
import Header from './Header';
import User from './User';
import Loading from './Loading';

export class Dashboard extends Component {
  renderUsers = () => {
    if (this.props.isLoading) {
      return <Loading />;
    }
    const users = this.props.users.map((user) => {
      const firstName = user.firstName.replace(this.props.filterText, `<span class="hl">${this.props.filterText}</span>`);
      const lastName = user.lastName.replace(this.props.filterText, `<span class="hl">${this.props.filterText}</span>`);

      const userData = {
        firstName,
        lastName,
        image: user.image,
        username: user.username,
        cell: user.cell,
      };

      return <User key={user.username} userData={userData} />;
    });
    return users;
  }

  render() {
    return (
      <div>
        <Header />
        <div className="container">
          <div className="row users pt-5">
            { this.renderUsers() }
          </div>
        </div>
      </div>
    );
  }
}

Dashboard.propTypes = {
  isLoading: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  users: filterUsers(state.users, state.filters),
  filterText: state.filters.text,
  isLoading: state.usersAreLoading,
});

export default connect(mapStateToProps)(Dashboard);
