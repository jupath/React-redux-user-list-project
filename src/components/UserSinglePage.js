import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { Button } from 'reactstrap';
import UserSinglePageField from './UserSinglePageField';

export class UserSinglePage extends Component {

  render() {

    const dob = moment(this.props.user.dob).format('Do MMMM, YYYY');

    return (
      <div className="container">
        <div className="px-lg-6 mt-6 mb-4">
          <div className="row user-single p-3">
            <div className="pr-3">
              <img src={this.props.user.image} />
            </div>
            <div className="user-single__details">
              <h2 className="text-capitalize font-weight-bold">{this.props.user.firstName} {this.props.user.lastName}</h2>
              <div className="user-single__details__field">
                <div className="font-weight-bold pr-2">Email:</div>
                <UserSinglePageField data={this.props.user.email} field={'email'} username={this.props.user.username} isEditable={true} />
              </div>
              <div className="user-single__details__field">
                <div className="font-weight-bold pr-2">Cell:</div>
                <UserSinglePageField data={this.props.user.cell} field={'cell'} username={this.props.user.username} isEditable={true}/>
              </div>
              <div className="user-single__details__field text-capitalize">
                <div className="font-weight-bold pr-2">Address:</div>
                <UserSinglePageField data={this.props.user.address} field={'address'} username={this.props.user.username} isEditable={true} />
              </div>
              <div className="user-single__details__field">
                <div className="font-weight-bold pr-2">Day of birth:</div>
                <UserSinglePageField data={dob} field={'dob'} username={this.props.user.username} isEditable={false} />
              </div>
              <Button color="success" className="mt-3" onClick={() => this.props.history.push('/')}>Back to dashboard</Button>
            </div>
          </div>
        </div>
      </div>
    );
  };
}

const mapStateToProps = (state, props) => ({
  user: state.users.find((user) => user.username === props.match.params.username)
});

export default connect(mapStateToProps)(UserSinglePage);