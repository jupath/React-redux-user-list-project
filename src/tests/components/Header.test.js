import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

test('should render Header correctly when orderBy is undefined', () => {
  const wrapper = shallow(<Header />);
  expect(wrapper).toMatchSnapshot();
});

test('should render Header correctly when orderBy is defined', () => {
  const wrapper = shallow(<Header orderBy={'asc'} />);
  expect(wrapper).toMatchSnapshot();
});

test('should handle orderBy button click', () => {
  const setOrderBySpy = jest.fn();
  const wrapper = shallow(<Header setOrderBy={setOrderBySpy} />);
  wrapper.find('#ascButton').simulate('click');
  expect(setOrderBySpy).toHaveBeenLastCalledWith('asc');
});

test('should handle setText correctly', () => {
  const setTextSpy = jest.fn();
  const value = 'test';
  const wrapper = shallow(<Header setText={setTextSpy} />);
  wrapper.find('Input').simulate('change', {
    target: { value }
  });
  expect(setTextSpy).toHaveBeenLastCalledWith(value);
});