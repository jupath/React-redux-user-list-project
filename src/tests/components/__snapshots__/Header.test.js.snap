// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly when orderBy is defined 1`] = `
<header
  className="header"
>
  <div
    className="container text-center py-4"
  >
    <h1>
      User list built with React and Redux
    </h1>
    <p>
      This app uses async-await to fetch users from 
      <a
        href="https://randomuser.me/"
        target="_blank"
      >
        randomuser.me
      </a>
       API
    </p>
    <div
      className="px-lg-7 pb-4"
    >
      <Input
        className="header__searchform"
        onChange={[Function]}
        placeholder="Search by name..."
        tag="p"
        type="text"
      />
    </div>
    <div>
      <span>
        Order list by name: 
      </span>
      <Button
        className="mr-1"
        color="success"
        disabled={true}
        id="ascButton"
        onClick={[Function]}
        tag="button"
      >
        Asc
      </Button>
      <Button
        className="mr-1"
        color="success"
        disabled={false}
        id="descButton"
        onClick={[Function]}
        tag="button"
      >
        Desc
      </Button>
      <Button
        color="success"
        disabled={false}
        id="defaultButton"
        onClick={[Function]}
        tag="button"
      >
        Default
      </Button>
    </div>
  </div>
</header>
`;

exports[`should render Header correctly when orderBy is undefined 1`] = `
<header
  className="header"
>
  <div
    className="container text-center py-4"
  >
    <h1>
      User list built with React and Redux
    </h1>
    <p>
      This app uses async-await to fetch users from 
      <a
        href="https://randomuser.me/"
        target="_blank"
      >
        randomuser.me
      </a>
       API
    </p>
    <div
      className="px-lg-7 pb-4"
    >
      <Input
        className="header__searchform"
        onChange={[Function]}
        placeholder="Search by name..."
        tag="p"
        type="text"
      />
    </div>
    <div>
      <span>
        Order list by name: 
      </span>
      <Button
        className="mr-1"
        color="success"
        disabled={false}
        id="ascButton"
        onClick={[Function]}
        tag="button"
      >
        Asc
      </Button>
      <Button
        className="mr-1"
        color="success"
        disabled={false}
        id="descButton"
        onClick={[Function]}
        tag="button"
      >
        Desc
      </Button>
      <Button
        color="success"
        disabled={true}
        id="defaultButton"
        onClick={[Function]}
        tag="button"
      >
        Default
      </Button>
    </div>
  </div>
</header>
`;
